### ISSUE TITLE
[Blazor WASM] InputSelect: binding no longer works with negative numbers

### ISSUE THREAD

**Issue by MichelJansson** [dotnet/aspnetcore#42064]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Data binding to an html input `<input @bind="xyz">` OR `<InputSelect @bind-Value="xyz">` no longer works with negative numbers (`-1`). Only an empty select is rendered (value of select in inspector is empty string)

Unsure when it stopped working, but i'm pretty sure this worked in 6.0.0 as I have components explicitly built that uses negative numbers (to support nullable enum and and optional grpc enum)

### Expected Behavior

Binding to write the negative number to the select element, so the negative values can be displayed and selected.

### Steps To Reproduce

```razor
<select @bind="@myValue">
    <option value="-1">-1</option>
    <option value="0">0</option>
    <option value="1">1</option>
</select>

@code {
    private int myValue = -1;
}
```

1. Select shows up as empty on load
2. Selecting positive number works as expected
3. Selecting a negative number, the select again becomes emtpy.

### .NET Version

6.0.300

### Anything else?

ASP.NET Core 6.0.5
Visual Studio 17.2.3

```
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.300
 Commit:    8473146e7d

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.300\

Host (useful for support):
  Version: 6.0.5
  Commit:  70ae3df4a6

.NET SDKs installed:
  3.1.302 [C:\Program Files\dotnet\sdk]
  3.1.401 [C:\Program Files\dotnet\sdk]
  3.1.418 [C:\Program Files\dotnet\sdk]
  5.0.101 [C:\Program Files\dotnet\sdk]
  5.0.404 [C:\Program Files\dotnet\sdk]
  5.0.407 [C:\Program Files\dotnet\sdk]
  6.0.300 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.24 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by danroth27** [comment]
I'm not able to reproduce this: https://blazorrepl.telerik.com/QQkAYhPA46GlQcOi22

@MichelJansson Can you share a repro project?

**Comment by MichelJansson** [comment]
> I'm not able to reproduce this: https://blazorrepl.telerik.com/QQkAYhPA46GlQcOi22
> 
> @MichelJansson Can you share a repro project?

@danroth27 Thanks for looking into this! Hmm - the link you shared does not work for me (same issue as reported).
Expected to see "-1" in the select box, but it's blank.
![image](https://user-images.githubusercontent.com/29315027/172440211-276beb02-3256-4ed1-879d-17625c5fd6cb.png)

But since you said it works, I tested it with Edge (I guess that is what you used 😋) and it actually works there, so this must be Chrome-releated. 

The bug manifests it self Chrome Version 102.0.5005.63 (Official Build) (64-bit).

Is there any good way to debug bindings, so that I can help troubleshoot?

**Comment by danroth27** [comment]
> The bug manifests it self Chrome Version 102.0.5005.63 (Official Build) (64-bit).

Hmm...I'm not able to reproduce this in Chrome either:

![image](https://user-images.githubusercontent.com/1874516/172445812-32522db2-cab1-4e03-803a-f3bdda23f3e0.png)

**Comment by danroth27** [comment]
@MichelJansson Are there any errors in the browser dev console?

@javiercn @SteveSandersonMS Any suggestions on how to debug what's going on here?

**Comment by MichelJansson** [comment]
No errors in console.

I think I maybe managed to narrow it down somewhat by stepping the minified JS:
- It's after the linked line the select goes blank: https://github.com/dotnet/aspnetcore/blob/v6.0.5/src/Components/Web.JS/src/Rendering/BrowserRenderer.ts#L582
- If I, while at the break-point, edit and reset the value to be written (just writing "-1" again) it works as expected.

- At closer inspection of the string that blazor tries to set for int -1 is using a different character for the minus sign:
  It tries to write  `"−1"` (hyphen) while it should be `"-1"` (hyphen-minus)

-----

Something way back in my head tells me I've seen a similar issue with javascript in the past.
Can it be culture setting regarding number formatting? I saw somewhere in the same JS file that `JSON.parse()` was used to convert the value?

**Comment by jessegood** [comment]
Seems it may be related to using ICU?
See https://github.com/dotnet/runtime/issues/47524

**Comment by MichelJansson** [comment]
> Seems it may be related to using ICU? See [dotnet/runtime#47524](https://github.com/dotnet/runtime/issues/47524)

Thanks! Although I don't fully understand that issue, the symptom seems to be the same, and I too am using Swedish locale.

------

As a workaround for the pressing issue with my select's, I currently use actual int's for the values on the options that then gets converted in the same way as the bind:
```razor
<select @bind="@myValue">
    <option value="@(-1)">-1</option>
    <option value="@(0)">0</option>
    <option value="@(1)">1</option>
</select>

@code {
    private int myValue = -1;
}
```

**Comment by TanayParikh** [comment]
@MichelJansson can you please confirm if you're still experiencing this issue in `.NET 7`? I just tried it out, and haven't been able to repro:

<img width="800" alt="image" src="https://user-images.githubusercontent.com/14852843/183319574-980043cf-cc39-40a4-8718-d159d5d85441.png">

For additional context, as @jessegood suggested above, this seems to be locale specific. I repro'd using `sv`, I wasn't able to repro using `en-US`.

**Comment by TanayParikh** [comment]
Unable to get any kind of repro in .NET 7.

<img width="736" alt="image" src="https://user-images.githubusercontent.com/14852843/183514908-6a0d6b4d-128a-4ce0-951f-435542f5f393.png">

Works for both the unicode minus sign, as well as floats. Blazor Server & WebAssembly.

vs. in .NET 6:

<img width="314" alt="image" src="https://user-images.githubusercontent.com/14852843/183515005-2d0d3a8b-33af-45e7-a4b3-642150a0aecb.png">

@MackinnonBuck had called out:

https://github.com/dotnet/aspnetcore/blob/51ccce6d743dddf995b3661f07b0e1990b68dc7d/src/Components/Web/src/Forms/InputExtensions.cs#L35

https://github.com/dotnet/aspnetcore/blob/51ccce6d743dddf995b3661f07b0e1990b68dc7d/src/Components/Web/src/Forms/InputSelect.cs#L84

however, even without changing the culture to `CultureInfo.InvariantCulture` we're seeing the expected behavior. I believe to prevent inadvertently introducing unintended behavior we may wish to close out this issue for now, until we get feedback on a particular sample/culture where things are _not_ working.

cc/ @SteveSandersonMS

<details>
<summary>Sample Code</summary>

```razor

<select @bind="@negativeVal">
    <option value="-1">-1</option>
    <option value="0">0</option>
    <option value="1">1</option>
</select>

@code {
    private int negativeVal = -1;
}

Negative Value is @negativeVal


<select @bind="@myValue">
    <option value="0.23">0.23</option>
    <option value="0.24">0.24</option>
    <option value="0.25">0.25</option>
</select>

Value is @myValue

@code {
    private double myValue = 0.23;
}
```

</details>

**Comment by MichelJansson** [comment]
@TanayParikh, thanks for looking into this. I have yet to install .NET7, will have to get back to you later.

**Comment by TanayParikh** [comment]
> Unable to get any kind of repro in .NET 7.
> vs. in .NET 6:

I was using the `aspnetcore` repo for `.NET7` repro, and https://blazorrepl.telerik.com/ for the `.NET6` repro.

I just tried this out using the `aspnetcore` repo on the `release/6.0` branch, and not able to repro it their either:

<img width="379" alt="image" src="https://user-images.githubusercontent.com/14852843/183518345-38cf4ed4-a916-43f6-b109-f356dae28ed2.png">

So it seems the telerik is doing some type of special processing (unclear what, `Startup.cs` is empty).

**Comment by SteveSandersonMS** [comment]
@TanayParikh Thanks for continuing the investigation!

I agree it looks like there's **no further problem relating to the minus character in Swedish culture**, since that now seems to be accepted for culture-specific parsing.

However, the underlying question about whether we should be doing culture-specific parsing at all for `<select>` and `<input type="radio">` is still worth us thinking through clearly. To illustrate the difference, I've created a test case at https://blazorrepl.telerik.com/wwEsuXuN56lSQqKn02. That's running on .NET 6, but behaves the same on 7 too.

The behavior is:

 * If your browser language has similar-enough number formatting to "invariant culture" (e.g., `en-US` or `en-GB`), then all the dropdowns work
 * But if your browser is set to French, for example, then you can't pick *three point five* from the top two dropdowns, but all the other dropdown choices work
   * The reason is that `@((3.5).ToString(CultureInfo.InvariantCulture))` renders as the string `3.5`, but when we try to parse that culture-sensitively, it just doesn't parse. Similarly, if the `myValue` field already holds the number three-point-five, then it doesn't match any of the `<option>` values since it gets formatted as `3,5` and no option has a value with that string.
   * The values *one* and *thirty five* both do work though, since their French-formatted value happens to be equivalent to the invariant-formatted value

Yesterday I thought it was quite clear we should be doing culture-invariant parsing/formatting for `<select>` and `<input type="radio">` values, but on further consideration I'm not so sure. The unavoidable truth is that those elements have *string* values, not numeric values. Whether you want to parse as culture-invariant or specific depends on whether you previously formatted the choices as culture-invariant or specific, and that's not something that Blazor controls, unless we choose to change how it works.

For example, if your `.razor` source code includes:

```razor
<select @bind="someValue">
    <option value="-1.5">minus one point five</option>
    <option value="1.5">one point five</option>
</select>
```

... then those two `<option>` elements are literal HTML text, and Blazor doesn't interfere with their contents at all. It's up to the developer to type out the values as strings in the right culture.

Whereas if your source code includes the same except:

```razor
    <option value="@(-1.5)">minus one point five</option>
```

... now Blazor *does* format the C# number literal as a string for you, and our default is to format all output as culture-specific, since it's usually used for display to the user. We could look into keeping track of the context of where a given value was being outputted, and varying the display format, but that's a whole new complexity we don't deal with today.

So overall, our existing culture-sensitive parsing for `<select>` and `<input type="radio">`:

 * Is good if the developer types string literals in the same culture as their end-users (fairly common)
 * Is bad if that same developer has some end-users with a different culture
 * Is good if the developer emits values using `@someValue`, because that will dynamically vary the format to match each user
 * Is bad if the developer emits values using `@(someValue.ToString(CultureInfo.InvariantCulture))`, because our culture-specific parsing will fail if the user isn't in an invariant-like culture

I don't think we really can change it to be culture-invariant by default, because then we'd require developers to always either:

 * Type out string literals formatted in invariant culture, regardless of their own culture, which they won't remember to do
 * Or, always use `.ToString(CultureInfo.InvariantCulture)`, which they also won't remember to do

One person in the meeting who seemed to have a clear preference yesterday was @Eilon - do you have any feedback on this?

Anyone else?

**Comment by Eilon** [comment]
I think that logically the `value` of an `<option>` is culture-invariant, because those values are strictly server-supplied, and servers logically operate as "invariant culture." However, because "things are complicated" and "the server doesn't really know what the client is doing" and "it's a breaking change" I understand entirely that it's not something we could really change. (Here in Blazor WASM there's no "server" but you can take it to mean the "logic behind the page".)

Historically one way we've suggested for people to work around this is to always bind values to strings, and then in the class being bound to, have some logic that parses the string however you want (whatever culture you want, trim spaces, do conversions, etc.).

**Comment by danroth27** [comment]
Yeah, I don't really see an argument for those literal strings to be treated as culture-invariant. It seems more likely that they are culture unknown, but most likely in the culture of whoever typed them.

**Comment by MichelJansson** [comment]
Thanks for all your brain power! Personally I would have expected the `value` attribute to be invariant-culture and agree with @Eilon.
But as for the workaround to use strings and having to do manual conversion logic for each field is a big pain. I think the suggested workaround should rather be as we have discussed here to use `@(-1)`.

@TanayParikh, for completeness, I just installed .NET7-p7 and still could repro using my original snippet.
